File: CPP/ByRound/1704/1704_E_Count_Seconds.cpp
// Problem: CF 1704 E - Count Seconds
// https://codeforces.com/contest/1704/problem/E

/*
 * Problem: Count Seconds
 * 
 * Purpose: 
 *   Given a directed acyclic graph (DAG) with node values, simulate a process
 *   where each second, all nodes with positive values decrement by 1, and
 *   their outgoing edges propagate 1 to the next nodes. Return the time when
 *   all values become 0.
 * 
 * Algorithm:
 *   - Simulation with topological sorting
 *   - For small steps, simulate the process directly
 *   - For large steps, compute the final contribution using topological order
 * 
 * Time Complexity:
 *   O(n^2) per test case in worst case due to simulation, but optimized via
 *   early termination when all nodes reach 0.
 * 
 * Space Complexity:
 *   O(n + m) for adjacency list and auxiliary arrays.
 * 
 * Techniques:
 *   - Simulation with topological sorting
 *   - Modular arithmetic for large results
 * 
 * Note:
 *   The first part simulates step-by-step until all values are 0.
 *   The second part calculates remaining time with topological order.
 * 
 */

#include <algorithm>
#include <iostream>
#include <vector>

#define ll long long
#define mod 998244353
using namespace std;

ll a[1010], na[1010], rd[1010], vis[1010];
vector<int> e[1010];

int main() {
  int T;
  cin >> T;
  while (T--) {
    int n, m, i, j, flg = 0;
    cin >> n >> m;
    for (i = 1; i <= n; i++) {
      cin >> a[i];
      e[i].clear();
      rd[i] = 0;
      vis[i] = 0;
    }
    for (i = 1; i <= m; i++) {
      int u, v;
      cin >> u >> v;
      e[u].push_back(v);
      rd[v]++;
    }
    ll ans = 0;
    for (i = 1; i <= n + 2; i++) {
      int f = 0;
      for (j = 1; j <= n; j++) {
        na[j] = a[j];
        if (a[j])
          f = 1;
      }
      if (!f) {
        flg = 1;
        break;
      }
      for (j = 1; j <= n; j++)
        if (a[j]) {
          na[j]--;
          f = 1;
          for (auto p : e[j])
            na[p]++;
        }
      for (j = 1; j <= n; j++)
        a[j] = na[j];
      ans++;
    }
    if (flg) {
      cout << ans % mod << endl;
      continue;
    }
    ll res = 0, s = 0;
    for (i = 1; i <= n; i++) {
      int p = 0;
      for (j = 1; j <= n; j++)
        if (!vis[j] && rd[j] == 0) {
          p = j;
          break;
        }
      vis[p] = 1;
      if (i == n) {
        res = a[p];
        break;
      }
      for (auto q : e[p])
        (a[q] += a[p]) %= mod, rd[q]--;
    }
    cout << (ans + res) % mod << endl;
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/E/1704_E_Count_Seconds.py
# Problem: CF 1704 E - Count Seconds
# https://codeforces.com/contest/1704/problem/E

"""
Problem: Count Seconds
Algorithm: Simulation with Graph Traversal
Time Complexity: O(n * (n + m)) where n is number of nodes and m is number of edges
Space Complexity: O(n + m) for storing graph and auxiliary arrays

This solution simulates the process described in the problem:
1. For each second, we find all nodes with positive values (S)
2. For each node in S:
   - Decrement its value by 1
   - Distribute 1 to all its neighbors
3. We track when all values become 0
4. Special handling for the node with no outgoing edges to ensure correct simulation

The simulation uses a simplified approach where each step updates the array based on the current state.
"""

T = int(input())
while T:
    T -= 1
    n, m = list(map(int, input().split()))
    arr = list(map(int, input().split()))
    g = [[] for _ in range(n)]
    deg = [0] * n
    out = [0] * n
    for _ in range(m):
        x, y = list(map(int, input().split()))
        x -= 1
        y -= 1
        g[x].append(y)
        deg[y] += 1
        out[x] += 1

    index = -1
    for i in range(n):
        if out[i] == 0:
            index = i
            break

    t = 0
    for _ in range(n): # At most n steps needed
        if not any(arr): # If all values are 0
            break
        t += 1
        nxt = [0] * n
        for x in range(n):
            for y in g[x]: # For each neighbor of x
                nxt[y] += arr[x] # Add value from x to y
        nxt[index] += max(0, arr[index] - 1) # Handle the special node
        arr = nxt

    print((t + arr[index]) % 998244353)


# https://github.com/VaHiX/CodeForces/